src/psyclone/tests/gocean1p0_opencl_test.py

```diff
@@ -264,70 +359,126 @@ def test_grid_proprty(kernel_outputdir):
 
-    otrans = OCLTrans()
-    otrans.apply(sched)
-    generated_code = str(psy.gen).lower()
+    # Check that the device buffer initialisation routine has been generated
+    expected = '''\
+    subroutine initialise_device_buffer(field)
+      use fortcl, only: create_rw_buffer
+      use field_mod
+      type(r2d_field), intent(inout), target :: field
+      integer(kind=c_size_t) size_in_bytes
 
-    # The array size expression always uses the same field, in this case p_fld
-    size_expression = "int(p_fld%grid%nx*p_fld%grid%ny, 8)"
-
-    # For each of the invoke fields, add a conditional block that:
-    # - Creates a OpenCL rw buffer.
-    # - Writes the field data into the buffer.
-    # - Marks data_on_device flag as true.
-    # - Points the read_from_device_f attribute to the read_from_device
-    # local function.
-    # - Blocks OpenCL Queue until the data copy has finished.
-    single_invoke_fields = ["cu_fld", "p_fld", "u_fld"]
-    for field in single_invoke_fields:
-        expected = (
-            "      if (.not. {0}%data_on_device) then\n"
-            "        size_in_bytes = " + size_expression +
-            "*c_sizeof({0}%data(1,1))\n"
-            "        ! create buffer on device\n"
-            "        {0}%device_ptr = create_rw_buffer(size_in_bytes)\n"
-            "        ierr = clenqueuewritebuffer(cmd_queues(1), "
-            "{0}%device_ptr, cl_true, 0_8, size_in_bytes, "
-            "c_loc({0}%data), 0, c_null_ptr, "
-            "c_loc(write_event))\n"
-            "        {0}%data_on_device = .true.\n"
-            "        {0}%read_from_device_f => read_from_device\n"
-            "        ! block until data copies have finished\n"
-            "        ierr = clfinish(cmd_queues(1))\n"
-            "      end if\n").format(field)
-        assert expected in generated_code
-
-    # Check that the read_from_device routine has also been generated.
-    expected = (
-        "    subroutine read_from_device(from, to, nx, ny, width)\n"
-        "      use iso_c_binding, only: c_intptr_t\n"
-        "      use fortcl, only: read_buffer\n"
-        "      integer(kind=c_intptr_t), intent(in) :: from\n"
-        "      real(kind=go_wp), intent(inout), dimension(:,:) :: to\n"
-        "      integer, intent(in) :: nx, ny, width\n"
-        "      call read_buffer(from, to, int(width*ny, kind=8))\n"
-        "    end subroutine read_from_device\n")
+      if (.not.field%data_on_device) then
+        size_in_bytes = int(field%grid%nx * field%grid%ny, 8) * \
+c_sizeof(field%data(1,1))
+        field%device_ptr = transfer(create_rw_buffer(size_in_bytes), \
+field%device_ptr)
+        field%data_on_device = .true.
+        field%read_from_device_f => read_from_device
+        field%write_to_device_f => write_to_device
+      end if
+
+    end subroutine initialise_device_buffer'''
+    print(generated_code)
     assert expected in generated_code
+    # We don't need a Compilation test because it is the same than
```

Comment:
I think we should have one anyway as a defensive measure against future cutting/pasting/moving of tests.

Reply:
Inserted Compilation assert